tests/Model/BSONIteratorTest.php

```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace MongoDB\Tests\Model;
+
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+use MongoDB\Model\BSONIterator;
+use MongoDB\Tests\TestCase;
+use stdClass;
+
+class BSONIteratorTest extends TestCase
+{
+    /**
+     * @dataProvider provideTypeMapOptionsAndExpectedDocuments
+     */
+    public function testValidValues(array $typeMap = null, $binaryString, array $expectedDocuments)
+    {
+        $bsonIt = new BSONIterator($binaryString, ['typeMap' => $typeMap]);
+
+        $results = iterator_to_array($bsonIt);
+
+        $this->assertSameDocuments($expectedDocuments, $results);
+    }
+
+    public function provideTypeMapOptionsAndExpectedDocuments()
+    {
+        return [
+            [
+                null,
+                implode(array_map(
+                    'MongoDB\BSON\fromPHP',
+                    [
+                        (object) ['_id' => 1, 'x' => (object) ['foo' => 'bar']],
+                        (object) ['_id' => 3, 'x' => (object) ['foo' => 'bar']],
```

Comment:
We can drop the object-casting from all of the `fromPHP()` inputs, unless you actually need an array like `[ 0 => 'foo' ]` to serialize as `{ "0": "foo" }` instead of `[ "foo" ]`.